### PR TITLE
fix: Updated shimmer to handle instrumenting named and default exports of CommonJS modules in ESM

### LINKED ISSUE #1821: ESM loader has multiple issues in v11
## Description

Using `--loader newrelic/esm-loader.mjs` in `v11` leads to several issues that did not exist in `v10`:
- Some packages are not properly instrumented. From my experiments, `ioredis`, `winston`, `amqplib`, `aws-sdk`, `prisma` are not properly instrumented (eg. no data posted), while `fastify` works (I'm getting `Fastify//*` transactions).
- It may depends on the "style" of the `import` statement. For instance, as detailed below, `winston` is properly instrumented in `v11` as long as it's imported like this `import winston from "winston"` and not like this `import * as winston from "winston"`. `v10` handles both cases properly.
- `v11` sometimes crashes silently. I experience this when importing `puppeteer` for instance. Potentially due to  https://github.com/DataDog/import-in-the-middle/issues/32

## Expected Behavior

No regression in ESM loader in `v11` wrt `v10`.

## Steps to Reproduce

This only illustrates the import style issue as this is the only one I could pinpoint.

```javascript
// v10/package.json
{
  "name": "v10",
  "type": "module",
  "dependencies": {
    "newrelic": "10.6.2",
    "winston": "3.11.0"
  }
}

// v10/import-all-as.js
import * as winston from 'winston';
const logger = winston.createLogger({ transports: new winston.transports.Console() });
logger.warn('V10: import * as');
await new Promise((resolve) => { setTimeout(() => resolve(), 5000); });

// v10/import-default.js
import winston from 'winston';
const logger = winston.createLogger({ transports: new winston.transports.Console() });
logger.warn('V10: default import');
await new Promise((resolve) => { setTimeout(() => resolve(), 5000); });

// v11/package.json
{
  "name": "v11",
  "type": "module",
  "dependencies": {
    "newrelic": "11.2.1",
    "winston": "3.11.0"
  }
}

// v11/import-all-as.js
import * as winston from 'winston';
const logger = winston.createLogger({ transports: new winston.transports.Console() });
logger.warn('V11: import * as');
await new Promise((resolve) => { setTimeout(() => resolve(), 5000); });

// v11/import-default.js
import winston from 'winston';
const logger = winston.createLogger({ transports: new winston.transports.Console() });
logger.warn('V11: default import');
await new Promise((resolve) => { setTimeout(() => resolve(), 5000); });
```

```bash
cd v10
NEW_RELIC_LICENSE_KEY=x NEW_RELIC_APP_NAME=x node --loader newrelic/esm-loader.mjs -r newrelic import-all-as.js
NEW_RELIC_LICENSE_KEY=x NEW_RELIC_APP_NAME=x node --loader newrelic/esm-loader.mjs -r newrelic import-default.js
cd ../v11
NEW_RELIC_LICENSE_KEY=x NEW_RELIC_APP_NAME=x node --loader newrelic/esm-loader.mjs -r newrelic import-all-as.js
NEW_RELIC_LICENSE_KEY=x NEW_RELIC_APP_NAME=x node --loader newrelic/esm-loader.mjs -r newrelic import-default.js
```
Results in dashboard:
newrelic-import-v11
No `V11: import * as` log entry.

## Your Environment

* Node 18
* MacOS

### PR DESCRIPTION
**Pull request by jsumners-nr**
This PR resolves #1821.

### CODE REVIEW

**Review by mrickard — APPROVED** [review]
LGTM, and much more readable. @bizob2828 should weigh in as well.

File: test/unit/util/codec.test.js [hunk 1]
```diff
@@ -22,13 +23,14 @@ test('codec', function (t) {
...
     t.test('should not error for circular payloads', function (t) {
-      const val = 'eJyrVkrLz1eyUkpKLFLSUcpPygKyo50zi5JLcxKLFOpilWoBuCkK6A=='
+      const val = '{"foo":"bar","obj":"[Circular ~]"}'
       const obj = { foo: 'bar' }
       obj.obj = obj
 
       codec.encode(obj, function (err, encoded) {
         t.error(err)
-        t.equal(encoded, val)
+        const decoded = zlib.inflateSync(Buffer.from(encoded, 'base64')).toString()
```
Comment (bizob2828) [inline test/unit/util/codec.test.js:32]: what does this have to do with the PR you opened?

File: lib/shimmer.js [hunk 2]
```diff
@@ -579,6 +576,60 @@ function instrumentPostLoad(agent, nodule, moduleName, resolvedName, esmResolver
...
+  // We have a CJS module wrapped by import-in-the-middle having been
+  // imported through ESM syntax. Due to the way CJS modules are parsed by
+  // ESM's import, we can have the same "export" attached to the `default`
+  // export and as a top-level named export. In order to shim things such
+  // that our users don't need to know to access `something.default.foo`
+  // when they have done `import * as something from 'something'`, we need
+  // to proxy the proxy in order to set our wrappers on both instances.
+  const noduleDefault = nodule.default
+  const origNodule = nodule
+  return new Proxy(
```
Comment (bizob2828) [inline lib/shimmer.js:608]: this is the root of the fix but the conventional commit is very vague and technically untrue.  This fixes all cjs modules and instrumenting the right thing depending on the import style correct?

File: lib/shimmer.js [hunk 3]
```diff
@@ -579,6 +576,60 @@ function instrumentPostLoad(agent, nodule, moduleName, resolvedName, esmResolver
...
+    { origNodule, noduleDefault },
+    {
+      get(target, name) {
+        if (name === Symbol.for('nr-esm-proxy')) {
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/shimmer.js:612]: oh, this is telling instrumentation it's getting proxied. You may want to put the string name in `lib/symbols though`

File: lib/instrumentation/pg.js [hunk 4]
```diff
@@ -19,6 +19,7 @@ function getQuery(shim, original, name, args) {
   return statement
 }
 
+// eslint-disable-next-line sonarjs/cognitive-complexity
```
Comment (bizob2828) [inline lib/instrumentation/pg.js:22]: we try not to add these overrides, they only exist for when we introduce the rule and are trying to pay down. i also noticed it exists in openai and i missed that.  I'll file a ticket to address

**Review by bizob2828 — CHANGES_REQUESTED** [review]
overall looks good. I'm not sure of the impact of proxying proxies but appears to work with our versioned tests.  I do have a few comments and small things that need addressed

File: lib/shimmer.js (hunk 2 above)
Comment (jsumners-nr) [reply lib/shimmer.js:608]: Yes, that is an accurate assessment. I must admit that I'm not well versed in conventional commit. What message do you think would be more appropriate?

File: bin/macos-firewall.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (jsumners-nr) [reply bin/macos-firewall.sh:1]: Added some docs in commit https://github.com/newrelic/node-newrelic/pull/1894/commits/b2ed29053716b6ad8e9bac3ceb243770e3ed4065.

**Comment by jsumners-nr** [comment]
> I'm not sure of the impact of proxying proxies but appears to work with our versioned tests.

Frankly, it's a terrible idea. But we need to double up the decoration for CJS imported as ESM, otherwise we have to educate users that the imports do not work as they expect. I do think the IITM team is open to moving this work into the IITM module. So I'm going to work up a patch over there that will do the same thing and see if we can then simplify here. But I am convinced we need the symbol added in this implementation and I do not think they are open to that bit.

File: lib/instrumentation/pg.js (hunk 4 above)
Comment (jsumners-nr) [reply lib/instrumentation/pg.js]: :-/ and doing it again seems to have worked. I must've had an autocompletion error in my first attempt.

**Review by bizob2828 — APPROVED** [review]
